pkg/builtin/multi/substring.go

```diff
@@ -0,0 +1,644 @@
+// Copyright 2021 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multi
+
+import (
+	"fmt"
+
+	"github.com/matrixorigin/matrixone/pkg/builtin"
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/sql/colexec/extend"
+	"github.com/matrixorigin/matrixone/pkg/sql/colexec/extend/overload"
+	"github.com/matrixorigin/matrixone/pkg/vectorize/substring"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+)
+
+//declare function parameter types and return types.
+var argAndrets = []argsAndRet{
```

Comment:
gee, in a situation with this amount of parameter type permutations, I think we can simply add a cast rule to cast the parameter type to a particular one.

Reply:
I think so, but I need to confirm whether the matching rules of these parameters and return value types are related to the frontend parser

Reply:
perhaps use a loop would simplify this code